File: src/test/java/org/apache/rocketmq/broker/server/TransactionMessageTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.broker.server;

import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.concurrent.Callable;
import java.util.concurrent.atomic.AtomicInteger;
import org.apache.rocketmq.client.apis.consumer.FilterExpression;
import org.apache.rocketmq.client.apis.message.Message;
import org.apache.rocketmq.client.apis.message.MessageView;
import org.apache.rocketmq.client.apis.producer.TransactionChecker;
import org.apache.rocketmq.client.apis.producer.TransactionResolution;
import org.apache.rocketmq.client.rmq.RMQNormalConsumer;
import org.apache.rocketmq.client.rmq.RMQNormalProducer;
import org.apache.rocketmq.client.trans.LocalTransactionCheckerImpl;
import org.apache.rocketmq.common.attribute.TopicMessageType;
import org.apache.rocketmq.enums.TESTSET;
import org.apache.rocketmq.factory.ConsumerFactory;
import org.apache.rocketmq.factory.MessageFactory;
import org.apache.rocketmq.factory.ProducerFactory;
import org.apache.rocketmq.frame.BaseOperate;
import org.apache.rocketmq.listener.rmq.RMQNormalListener;
import org.apache.rocketmq.util.NameUtils;
import org.apache.rocketmq.util.RandomUtils;
import org.apache.rocketmq.util.TestUtils;
import org.apache.rocketmq.util.VerifyUtils;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

@Tag(TESTSET.TRANSACTION)
@Tag(TESTSET.SMOKE)
public class TransactionMessageTest extends BaseOperate {
    private final Logger log = LoggerFactory.getLogger(TransactionMessageTest.class);
    private String tag;
    private final static int SEND_NUM = 10;
    private RMQNormalProducer producer;
    private RMQNormalConsumer pushConsumer;
    private RMQNormalConsumer simpleConsumer;

    @BeforeEach
    public void setUp() {
        tag = NameUtils.getRandomTagName();
    }

    @AfterEach
    public void tearDown() {
        if (producer != null) {
            producer.close();
        }
        if (pushConsumer != null) {
            pushConsumer.close();
        }
        if (simpleConsumer != null) {
            simpleConsumer.close();
        }
    }

    @Test
    @DisplayName("Send 10 transaction messages and synchronously commit the transaction (Checker performs rollback), expecting those 10 messages to be consumed via PushConsumer")
    public void testTrans_SendCommit_PushConsume() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();

        String topic = getTopic(TopicMessageType.TRANSACTION.getValue(), methodName);
        String groupId = getGroupId(methodName);

        pushConsumer = ConsumerFactory.getRMQPushConsumer(account, topic, groupId, new FilterExpression(tag), new RMQNormalListener());
        simpleConsumer = ConsumerFactory.getRMQSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer.getSimpleConsumer());

        producer = ProducerFactory.getRMQTransProducer(account, topic, new LocalTransactionCheckerImpl(TransactionResolution.ROLLBACK));
        Assertions.assertNotNull(producer);
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, RandomUtils.getStringByUUID());
            producer.sendTrans(message, true);
        }
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        VerifyUtils.verifyNormalMessage(producer.getEnqueueMessages(), pushConsumer.getListener().getDequeueMessages());
    }

    @Test
    @DisplayName("Send 10 transaction messages and rollback directly (Checker does commit), expecting that these 10 messages cannot be consumed by PushConsumer")
    public void testTrans_SendRollback_PushConsume() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(TopicMessageType.TRANSACTION.getValue(), methodName);
        String groupId = getGroupId(methodName);

        pushConsumer = ConsumerFactory.getRMQPushConsumer(account, topic, groupId, new FilterExpression(tag), new RMQNormalListener());
        simpleConsumer = ConsumerFactory.getRMQSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer.getSimpleConsumer());

        producer = ProducerFactory.getRMQTransProducer(account, topic, new LocalTransactionCheckerImpl(TransactionResolution.COMMIT));
        Assertions.assertNotNull(producer);
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, RandomUtils.getStringByUUID());
            producer.sendTrans(message, false);
        }
        //Wait for the callback, expecting not to commit the already rolled back message
        TestUtils.waitForSeconds(60);
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        Assertions.assertEquals(0, pushConsumer.getListener().getDequeueMessages().getDataSize());
    }

    @Test
    @DisplayName("Send 10 transaction messages and COMMIT the transaction by Checker (perform COMMIT), expecting the 10 messages to be consumed by PushConsumer")
    public void testTrans_SendCheckerCommit_PushConsume() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();

        String topic = getTopic(TopicMessageType.TRANSACTION.getValue(), methodName);
        String groupId = getGroupId(methodName);

        pushConsumer = ConsumerFactory.getRMQPushConsumer(account, topic, groupId, new FilterExpression(tag), new RMQNormalListener());
        simpleConsumer = ConsumerFactory.getRMQSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer.getSimpleConsumer());

        producer = ProducerFactory.getRMQTransProducer(account, topic, new LocalTransactionCheckerImpl(TransactionResolution.COMMIT));
        Assertions.assertNotNull(producer);
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, RandomUtils.getStringByUUID());
            producer.sendTrans(message, null);
        }
        //Wait for the callback to execute commit
        TestUtils.waitForSeconds(60);
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        VerifyUtils.verifyNormalMessage(producer.getEnqueueMessages(), pushConsumer.getListener().getDequeueMessages());
    }

    @Test
    @DisplayName("Send 10 transaction messages and roll back the transaction by Checker (performing ROLLBACK), expecting that the 10 messages will not be consumed by PushConsumer")
    public void testTrans_CheckerRollback() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();
        String topic = getTopic(TopicMessageType.TRANSACTION.getValue(), methodName);
        String groupId = getGroupId(methodName);

        pushConsumer = ConsumerFactory.getRMQPushConsumer(account, topic, groupId, new FilterExpression(tag), new RMQNormalListener());
        simpleConsumer = ConsumerFactory.getRMQSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer.getSimpleConsumer());

        producer = ProducerFactory.getRMQTransProducer(account, topic, new LocalTransactionCheckerImpl(TransactionResolution.ROLLBACK));
        Assertions.assertNotNull(producer);
        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, RandomUtils.getStringByUUID());
            producer.sendTrans(message, null);
        }
        //Wait for the rollback and execute rollback
        TestUtils.waitForSeconds(60);
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        Assertions.assertEquals(0, pushConsumer.getListener().getDequeueMessages().getDataSize());
    }

    @Test
    @DisplayName("Send 10 transactional messages and commit them by checking back (Checker commits for partial messages), and the expected committed messages can be consumed by PushConsumer")
    public void testTrans_SendCheckerPartionCommit() {
        String methodName = Thread.currentThread().getStackTrace()[1].getMethodName();

        String topic = getTopic(TopicMessageType.TRANSACTION.getValue(), methodName);
        String groupId = getGroupId(methodName);

        pushConsumer = ConsumerFactory.getRMQPushConsumer(account, topic, groupId, new FilterExpression(tag), new RMQNormalListener());
        simpleConsumer = ConsumerFactory.getRMQSimpleConsumer(account, topic, groupId, new FilterExpression(tag), Duration.ofSeconds(10));
        VerifyUtils.tryReceiveOnce(simpleConsumer.getSimpleConsumer());

        AtomicInteger commitMsgNum = new AtomicInteger(0);
        AtomicInteger rollbackMsgNum = new AtomicInteger(0);
        producer = ProducerFactory.getRMQTransProducer(account, topic, new TransactionChecker() {
            @Override
            public TransactionResolution check(MessageView messageView) {
                if (Integer.parseInt(StandardCharsets.UTF_8.decode(messageView.getBody()).toString()) % 2 == 0) {
                    commitMsgNum.getAndIncrement();
                    return TransactionResolution.COMMIT;
                } else {
                    rollbackMsgNum.getAndIncrement();
                    return TransactionResolution.ROLLBACK;
                }
            }
        });
        Assertions.assertNotNull(producer);

        for (int i = 0; i < SEND_NUM; i++) {
            Message message = MessageFactory.buildMessage(topic, tag, String.valueOf(i));
            producer.sendTrans(message, null);
        }
        await().atMost(90, SECONDS).until(new Callable<Boolean>() {
            @Override
            public Boolean call() {
                return rollbackMsgNum.get() == commitMsgNum.get() && commitMsgNum.get() == SEND_NUM / 2;
            }
        });
        //Wait for the rollback and execute commit/rollback
        TestUtils.waitForSeconds(60);
        Assertions.assertEquals(SEND_NUM, producer.getEnqueueMessages().getDataSize(), "send message failed");
        Assertions.assertEquals(5, pushConsumer.getListener().getDequeueMessages().getDataSize());
    }

}

File: cpp/rocketmq-client-cpp-tests/cpp4.x/test/server/TransactionMessageTest.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <atomic>
#include <chrono>
#include <cstddef>
#include <iostream>
#include <cassert>

#include <gtest/gtest.h>
#include <rocketmq/SendResult.h>
#include <spdlog/logger.h>
#include <string>

#include "enums/MessageType.h"
#include "frame/BaseOperate.h"
#include "resource/Resource.h"
#include "utils/NameUtils.h"
#include "utils/RandomUtils.h"
#include "utils/VerifyUtils.h"
#include "factory/ConsumerFactory.h"
#include "factory/ProducerFactory.h"
#include "factory/MessageFactory.h"
#include "common/MQTransactionListener.h"

extern std::shared_ptr<spdlog::logger> multi_logger;
extern std::shared_ptr<Resource> resource;

//Send 10 transaction messages and synchronously commit the transaction (Checker performs rollback), expecting those 10 messages to be consumed via PushConsumer
TEST(TransactionMessageTest, testTrans_SendCommit_PushConsume){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::TRANSACTION, "testTrans_SendCommit_PushConsume", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testTrans_SendCommit_PushConsume");
    std::string tag = NameUtils::getRandomTagName();

    auto pushConsumer = ConsumerFactory::getRMQPushConsumer(topic,group,tag,std::make_shared<RMQNormalListener>());

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));

    rocketmq::TransactionListener* listener = new CommitMQTransactionListener();
    auto transProducer = ProducerFactory::getRMQTransProducer(group,listener);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_NE(transProducer, nullptr);

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,RandomUtils::getStringByUUID());
        rocketmq::SendResult sendResult = transProducer->sendTrans(message,rocketmq::LocalTransactionState::COMMIT_MESSAGE);
        ASSERT_EQ(sendResult.getSendStatus(),rocketmq::SendStatus::SEND_OK);
    }

    ASSERT_EQ(SEND_NUM,transProducer->getEnqueueMessages()->getDataSize());

    ASSERT_TRUE(VerifyUtils::verifyNormalMessage(*(transProducer->getEnqueueMessages()),*(pushConsumer->getListener()->getDequeueMessages())));

    pushConsumer->shutdown();
    pullConsumer->shutdown();
    transProducer->shutdownTransaction();
}

//Send 10 transaction messages and rollback directly (Checker does commit), expecting that these 10 messages cannot be consumed by PushConsumer
TEST(TransactionMessageTest, testTrans_SendRollback_PushConsume){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::TRANSACTION, "testTrans_SendRollback_PushConsume", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testTrans_SendRollback_PushConsume");
    std::string tag = NameUtils::getRandomTagName();

    auto pushConsumer = ConsumerFactory::getRMQPushConsumer(topic,group,tag,std::make_shared<RMQNormalListener>());

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));
    pullConsumer->shutdown();

    rocketmq::TransactionListener* listener = new CommitMQTransactionListener();
    auto transProducer = ProducerFactory::getRMQTransProducer(group,listener);

    ASSERT_NE(transProducer, nullptr);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,RandomUtils::getStringByUUID());
        rocketmq::SendResult sendResult = transProducer->sendTrans(message,rocketmq::LocalTransactionState::ROLLBACK_MESSAGE);
        ASSERT_EQ(sendResult.getSendStatus(),rocketmq::SendStatus::SEND_OK);
    }
    
    std::this_thread::sleep_for(std::chrono::seconds(30));

    ASSERT_EQ(SEND_NUM,transProducer->getEnqueueMessages()->getDataSize());

    ASSERT_EQ(0,pushConsumer->getListener()->getDequeueMessages()->getDataSize());

    pushConsumer->shutdown();
    transProducer->shutdownTransaction();
}

//Send 10 transaction messages and COMMIT the transaction by Checker (perform COMMIT), expecting the 10 messages to be consumed by PushConsumer
TEST(TransactionMessageTest, testTrans_SendCheckerCommit_PushConsume){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::TRANSACTION, "testTrans_SendCheckerCommit_PushConsume", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testTrans_SendCheckerCommit_PushConsume");
    std::string tag = NameUtils::getRandomTagName();

    auto pushConsumer = ConsumerFactory::getRMQPushConsumer(topic,group,tag,std::make_shared<RMQNormalListener>());

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));
    pullConsumer->shutdown();

    rocketmq::TransactionListener* listener = new CommitMQTransactionListener();
    auto transProducer = ProducerFactory::getRMQTransProducer(group,listener);

    ASSERT_NE(transProducer, nullptr);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,RandomUtils::getStringByUUID());
        rocketmq::SendResult sendResult = transProducer->sendTrans(message,rocketmq::LocalTransactionState::UNKNOWN);
        ASSERT_EQ(sendResult.getSendStatus(),rocketmq::SendStatus::SEND_OK);
    }

    std::this_thread::sleep_for(std::chrono::seconds(30));

    ASSERT_EQ(SEND_NUM,transProducer->getEnqueueMessages()->getDataSize());

    ASSERT_TRUE(VerifyUtils::verifyNormalMessage(*(transProducer->getEnqueueMessages()),*(pushConsumer->getListener()->getDequeueMessages())));

    pushConsumer->shutdown();
    transProducer->shutdownTransaction();
}

//Send 10 transaction messages and roll back the transaction by Checker (performing ROLLBACK), expecting that the 10 messages will not be consumed by PushConsumer
TEST(TransactionMessageTest, testTrans_CheckerRollback){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::TRANSACTION, "testTrans_CheckerRollback", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testTrans_CheckerRollback");
    std::string tag = NameUtils::getRandomTagName();

    auto pushConsumer = ConsumerFactory::getRMQPushConsumer(topic,group,tag,std::make_shared<RMQNormalListener>());

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));
    pullConsumer->shutdown();

    rocketmq::TransactionListener* listener = new RollbackMQTransactionListener();
    auto transProducer = ProducerFactory::getRMQTransProducer(group,listener);

    ASSERT_NE(transProducer, nullptr);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,RandomUtils::getStringByUUID());
        rocketmq::SendResult sendResult = transProducer->sendTrans(message,rocketmq::LocalTransactionState::UNKNOWN);
        ASSERT_EQ(sendResult.getSendStatus(),rocketmq::SendStatus::SEND_OK);
    }

    std::this_thread::sleep_for(std::chrono::seconds(30));

    ASSERT_EQ(SEND_NUM,transProducer->getEnqueueMessages()->getDataSize());

    ASSERT_EQ(0,pushConsumer->getListener()->getDequeueMessages()->getDataSize());

    pushConsumer->shutdown();
    transProducer->shutdownTransaction();
}

TEST(TransactionMessageTest, testTrans_SendCheckerPartionCommit){
    int SEND_NUM = 10;
    std::string topic = getTopic(MessageType::TRANSACTION, "testTrans_SendCheckerPartionCommit", resource->getBrokerAddr(), resource->getNamesrv(),resource->getCluster());
    std::string group = getGroupId("testTrans_SendCheckerPartionCommit");
    std::string tag = NameUtils::getRandomTagName();

    auto pushConsumer = ConsumerFactory::getRMQPushConsumer(topic,group,tag,std::make_shared<RMQNormalListener>());

    auto pullConsumer = ConsumerFactory::getRMQPullConsumer(topic,group);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    ASSERT_TRUE(VerifyUtils::tryReceiveOnce(topic,tag,pullConsumer->getPullConsumer()));
    pullConsumer->shutdown();

    std::atomic<int> commitMsgNum(0);
    std::atomic<int> rollbackMsgNum(0);
    rocketmq::TransactionListener* listener = new UserdefinedMQTransactionListener(commitMsgNum,rollbackMsgNum);
    auto transProducer = ProducerFactory::getRMQTransProducer(group,listener);

    ASSERT_NE(transProducer, nullptr);

    std::this_thread::sleep_for(std::chrono::seconds(2));

    for(int i=0;i<SEND_NUM;i++){
        auto message = MessageFactory::buildMessage(topic,tag,std::to_string(i));
        rocketmq::SendResult sendResult = transProducer->sendTrans(message,rocketmq::LocalTransactionState::UNKNOWN);
        ASSERT_EQ(sendResult.getSendStatus(),rocketmq::SendStatus::SEND_OK);
    }

    int timeoutSeconds = 90;
    auto startTime = std::chrono::steady_clock::now();
    auto endTime = startTime + std::chrono::seconds(timeoutSeconds);
    while (commitMsgNum.load() != rollbackMsgNum.load() || commitMsgNum.load() != SEND_NUM/2) {
        if (std::chrono::steady_clock::now() >= endTime) {
            break;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    std::this_thread::sleep_for(std::chrono::seconds(30));

    ASSERT_EQ(SEND_NUM,transProducer->getEnqueueMessages()->getDataSize());

    ASSERT_EQ(SEND_NUM/2,pushConsumer->getListener()->getDequeueMessages()->getDataSize());

    pushConsumer->shutdown();
    transProducer->shutdownTransaction();
}